pkg/trtl/metrics/metrics.go

```diff
@@ -0,0 +1,239 @@
+package trtl
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/rs/zerolog/log"
+)
+
+var (
+	// Basic RPC Metrics
+	PmPuts       *prometheus.CounterVec   // count of trtl Puts per namespace
+	PmGets       *prometheus.CounterVec   // count of trtl Gets per namespace
+	PmDels       *prometheus.CounterVec   // count of trtl Deletes per namespace
+	PmIters      *prometheus.CounterVec   // count of trtl Iters per namespace
+	PmRPCLatency *prometheus.HistogramVec // the time it is taking for successful RPC calls to complete, labeled by RPC type, success, and failure
+	// PmObjects    *prometheus.CounterVec   // count of objects being managed by trtl, by namespace
+	// PmTombstones *prometheus.CounterVec   // count of tombstones per namespace; increases on delete, decrease on overwrite of tombstone
+
+	// Anti-Entropy Metrics
+	PmAESyncs         *prometheus.CounterVec   // count of anti entropy sessions per peer, per region, and by perspective (initiator/remote)
+	PmAESyncLatency   *prometheus.HistogramVec // duration of anti entropy sessions (initiator perspective), by peer
+	PmAEPhase1Latency *prometheus.HistogramVec // time phase 1 of anti-entropy is taking from the perspective of the initiator, by peer
+	PmAEPhase2Latency *prometheus.HistogramVec // time phase 2 of anti-entropy is taking from the perspective of the remote, by peer
+	PmAEVersions      *prometheus.HistogramVec // count of all observed versions, per peer and region
+	PmAEUpdates       *prometheus.HistogramVec // pushed objects during anti entropy, by peer and region
+	PmAERepairs       *prometheus.HistogramVec // pulled objects during anti entropy, by peer and region
+	PmAEStomps        *prometheus.CounterVec   // count of stomped versions, per peer and region
+	PmAESkips         *prometheus.CounterVec   // count of skipped versions, per peer and region
+
+)
+
+// A MetricsService manages Prometheus metrics
+type MetricsService struct {
+	srv *http.Server
+}
+
+func New() (*MetricsService, error) {
+	initMetrics()
```

Comment:
By adding `initMetrics` here, the metrics are unusable (they're nil, which will cause a panic if they're accessed) until the code creates a new `MetricsService`. This makes sense to me as it's the intended use case - what's the point of metrics without prometheus? (The alternative is to make `initMetrics` simply `init` - which will be called automatically when the package is imported the first time and there is no need to call `New`). However, I think that we need to document this in the comment above new, e.g.

```go
// New creates a metrics service and also initializes all of the prometheus metrics.
// Ensure that the trtl server creates the metrics service by calling New before any
// metrics are logged to Prometheus. Even in the case of tests, the metrics service
// must be created before the tests can be run. 
func New() (*MetricsService, error)

```